core/src/price_estimation/clients/apiclient.rs

```diff
@@ -1,79 +1,88 @@
-mod api;
-
-use super::{PriceSource, TokenData};
+use super::super::{PriceSource, TokenData};
 use crate::http::HttpFactory;
 use crate::models::TokenId;
 use anyhow::{anyhow, Context, Result};
-use api::{DexagApi, DexagHttpApi};
 use futures::{
     future::{self, BoxFuture, FutureExt as _},
     lock::Mutex,
 };
 use std::collections::HashMap;
 
-struct ApiTokens {
+pub trait Symbolic {
+    fn symbol(&self) -> &str;
+}
+
+pub trait Api {
+    type Token: Symbolic + Sync + Send;
+
+    fn get_token_list<'a>(&'a self) -> BoxFuture<'a, Result<Vec<Self::Token>>>;
+    /// Returns the price of one unit of `from` expressed in `to`.
+    /// For example `get_price("ETH", "DAI")` is ~220.
+    fn get_price<'a>(&'a self, from: &Self::Token, to: &Self::Token) -> BoxFuture<'a, Result<f64>>;
+}
+
+pub trait HttpConnecting: Sized {
+    /// Creates a new HTTP interface for connecting to the API
+    fn bind(http_factory: &HttpFactory) -> Result<Self>;
+}
+
+struct Tokens<T: Api> {
     // Maps uppercase Token::symbol to Token.
     // This is cached in the struct because we don't expect it to change often.
-    tokens: HashMap<String, api::Token>,
-    stable_coin: api::Token,
+    tokens: HashMap<String, T::Token>,
+    stable_coin: T::Token,
 }
 
-pub struct DexagClient<Api> {
-    api: Api,
+pub struct Client<T: Api> {
```

Comment:
nit: maybe the module can be called `generic_client` (as it works with multiple APIs, whereas  kraken requires a specific client)? No strong feelings (and I'm struggling to find a good reason beyond finding the word apiclient a bit weird).